packages/slate/src/models/text.js

```diff
@@ -241,82 +273,79 @@ class Text extends Record(DEFAULTS) {
   }
 
   /**
-   * Derive the leaves for a list of `characters`.
+   * Derive the leaves for a list of `decorations`.
    *
    * @param {Array|Void} decorations (optional)
    * @return {List<Leaf>}
    */
 
   getLeaves(decorations = []) {
-    const characters = this.getDecoratedCharacters(decorations)
-    let leaves = []
-
-    // PERF: cache previous values for faster lookup.
-    let prevChar
-    let prevLeaf
-
-    // If there are no characters, return one empty range.
-    if (characters.size == 0) {
-      leaves.push({})
-    } else {
-      // Otherwise, loop the characters and build the leaves...
-      characters.forEach((char, i) => {
-        const { marks, text } = char
-
-        // The first one can always just be created.
-        if (i == 0) {
-          prevChar = char
-          prevLeaf = { text, marks }
-          leaves.push(prevLeaf)
-          return
-        }
-
-        // Otherwise, compare the current and previous marks.
-        const prevMarks = prevChar.marks
-        const isSame = is(marks, prevMarks)
-
-        // If the marks are the same, add the text to the previous range.
-        if (isSame) {
-          prevChar = char
-          prevLeaf.text += text
+    let { leaves } = this
+    if (leaves.size === 0) return List.of(emptyLeaf)
+    if (!decorations || decorations.length === 0) return leaves
+    if (this.text.length === 0) return leaves
+    const { key } = this
+    decorations.forEach(range => {
+      const { startKey, endKey, startOffset, endOffset, marks } = range
+      const hasStart = startKey == key
+      const hasEnd = endKey == key
+      if (hasStart && hasEnd) {
+        const index = hasStart ? startOffset : 0
+        const length = hasEnd ? endOffset - index : this.text.length - index
+        if (length < 1) return
+        if (index >= this.text.length) return
+        if (index !== 0 || length < this.text.length) {
+          const [before, bundle] = Leaf.splitLeaves(this.leaves, index)
+          const [middle, after] = Leaf.splitLeaves(bundle, length)
+          leaves = before.concat(middle.map(x => x.addMarks(marks)), after)
           return
         }
-
-        // Otherwise, create a new range.
-        prevChar = char
-        prevLeaf = { text, marks }
-        leaves.push(prevLeaf)
-      }, [])
-    }
-
-    // PERF: convert the leaves to immutable objects after iterating.
-    leaves = new List(leaves.map(object => new Leaf(object)))
-
-    // Return the leaves.
-    return leaves
+      }
+      leaves = leaves.map(x => x.addMarks(marks))
+      return
```

Comment:
Unnecessary?

Reply:
Yes, the return is unnecessary. Fixed~ Thank you.